### PR TITLE
Do not store item pointers until all items are created

### PR DESCRIPTION
**Pull request by vigneshvg**
Calling avifMetaFindOrCreateItem() could invalidate all the existing item pointers that are being stored by the caller (since the function could resize the item array).

This patch fixes avifDecoderReset by storing the item indices instead of item pointers until all the items are either created or found.

### CODE REVIEW

**Comment by vigneshvg** [comment]
I have audited the code for all calls into `avifMetaFindOrCreateItem` and found that the only other function with this potential issue is `avifParseCondensedImageBox`. I will look into it and address that in a separate PR.

**Review by maryla-uc — APPROVED** [review]
Thanks for the fix!
To avoid this issue in the future maybe we could have a compile mode where avifArrayPush always resizes the array (only increases its capacity by 1) therefore alwyas invalidating pointers, that use it in tests?

**Comment by y-guyon** [comment]
Thanks for this PR, I encountered this issue multiple times in the past when refactoring the code, it is a classic blunder of `read.c` to me.  
I see that local pointers are used in most places for convenience, so risks remain despite passing indices instead of pointers as function arguments. Since pointers were more handy than indices, please consider using a fixed-size array or an array of pointers instead. I believe in C++ it would be a `std::vector<std::unique_ptr<avifDecoderItem>>` or a privately encapsulated `std::vector<avifDecoderItem>` with a public getter.

**Comment by vigneshvg** [comment]
> Thanks for this PR, I encountered this issue multiple times in the past when refactoring the code, it is a classic blunder of `read.c` to me. 

i agree.

> I see that local pointers are used in most places for convenience, so risks remain despite passing indices instead of pointers as function arguments.

I have audited the code for all uses of this function and there are only two places where we create new items on the go while potentially storing pointers to existing items. This PR fixes one of them (other one is `avifParseCondensedImageBox `). Every where else we only store read-only pointers which don't modify the items array. So it is safe to use them.

>  Since pointers were more handy than indices, please consider using a fixed-size array or an array of pointers instead. I believe in C++ it would be a `std::vector<std::unique_ptr<avifDecoderItem>>` or a privately encapsulated `std::vector<avifDecoderItem>` with a public getter.

I don't think a fixed size array can be used for items because the theoretical number of items can be too large. We could consider using a fixed size array of pointers but that seems like a bigger change and i will leave that to a separate PR.

File: src/read.c [hunk 1]
```diff
@@ -4712,6 +4726,14 @@ avifResult avifDecoderReset(avifDecoder * decoder)
         }
 #endif // AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP
 
+        // When searching for alpha and gainmap items, it is possible that the data->meta->items array is re-sized in calls to
+        // avifMetaFindOrCreateItem. So it is not safe to store pointers into that array until all the items are either found or
+        // created. From this point onwards, we do not create any more items, so it is safe to store the pointers.
```
[the remark this answers is not in the document]
Comment (vigneshvg) [reply src/read.c:4733]: Done.

File: src/read.c [hunk 2]
```diff
@@ -781,6 +781,8 @@ static avifResult avifCheckItemID(const char * boxFourcc, uint32_t itemID, avifD
     return AVIF_RESULT_OK;
 }
 
+// CAUTION: This function could potentially resize the meta->items array there by invalidating all existing pointers that are
```
Comment (wantehchang) [inline src/read.c:784]: Nit: there by => thereby

File: src/read.c [hunk 3]
```diff
@@ -4211,39 +4214,40 @@ static avifBool avifDecoderItemIsAlphaAux(avifDecoderItem * item, uint32_t color
     return auxCProp && isAlphaURN(auxCProp->u.auxC.auxType);
 }
 
-// Finds the alpha item whose parent item is colorItem and sets it in the alphaItem output parameter. Returns AVIF_RESULT_OK on
-// success. Note that *alphaItem can be NULL even if the return value is AVIF_RESULT_OK. If the colorItem is a grid and the alpha
-// item is represented as a set of auxl items to each color tile, then a fake item will be created and *isAlphaItemInInput will be
-// set to AVIF_FALSE. In this case, the alpha item merely exists to hold the locations of the alpha tile items. The data of this
-// item need not be read and the pixi property cannot be validated. Otherwise, *isAlphaItemInInput will be set to AVIF_TRUE when
-// *alphaItem is not NULL.
+// Finds the alpha item whose parent item is colorItem and sets its index in the alphaItemId output parameter. Returns
```
Comment (wantehchang) [inline src/read.c:4217]: Replace `colorItem` and `alphaItemId` with the new parameter names.

Perhaps `colorItem` can be replaced by "color item" (i.e., a description rather than a parameter name).

File: src/read.c [hunk 4]
```diff
@@ -4211,39 +4214,40 @@ static avifBool avifDecoderItemIsAlphaAux(avifDecoderItem * item, uint32_t color
...
+// AVIF_RESULT_OK on success. Note that *alphaItemIndex can be -1 even if the return value is AVIF_RESULT_OK. If the colorItem is
```
Comment (wantehchang) [inline src/read.c:4218]: `colorItem` needs to be replaced by `colorItemIndex` or "color item".

File: src/read.c [hunk 5]
```diff
@@ -4268,24 +4272,29 @@ static avifResult avifMetaFindAlphaItem(avifMeta * meta,
...
+    const avifResult result = avifMetaFindOrCreateItem(meta, maxItemID + 1, &alphaItem); // Create new empty item.
     if (result != AVIF_RESULT_OK) {
         avifFree(alphaItemIndices);
         *isAlphaItemInInput = AVIF_FALSE;
         return result;
     }
-    memcpy((*alphaItem)->type, "grid", 4); // Make it a grid and register alpha items as its tiles.
-    (*alphaItem)->width = colorItem->width;
-    (*alphaItem)->height = colorItem->height;
+    // avifMetaFindOrCreateItem() could invalidate all existing item pointers. So repopulate the colorItem pointer.
```
Comment (wantehchang) [inline src/read.c:4286]: Nit: "repopulate" implies filling out the members of a struct. Perhaps "reassign" or "reset"?

File: src/read.c [hunk 6]
```diff
@@ -4268,24 +4272,29 @@ static avifResult avifMetaFindAlphaItem(avifMeta * meta,
...
+    memcpy(alphaItem->type, "grid", 4); // Make it a grid and register alpha items as its tiles.
+    alphaItem->width = colorItem->width;
+    alphaItem->height = colorItem->height;
     for (uint32_t i = 0; i < alphaItemCount; ++i) {
         avifDecoderItem * item = &meta->items.item[alphaItemIndices[i]];
-        item->dimgForID = (*alphaItem)->id;
+        item->dimgForID = alphaItem->id;
     }
     avifFree(alphaItemIndices);
+    *alphaItemIndex = meta->items.count - 1;
```
Comment (wantehchang) [inline src/read.c:4297]: Nit/optional: It seems clearer to do this assignment immediately after the `avifMetaFindOrCreateItem()` call. We can probably wait until we know the call is successful, after line 4285.

File: src/read.c [hunk 7]
```diff
@@ -4342,41 +4351,49 @@ static avifResult avifDecoderDataFindToneMappedImageItem(const avifDecoderData *
...
+        avifDecoderDataFindToneMappedImageItem(data, &data->meta->items.item[colorItemIndex], &toneMappedImageItemTmp, &gainMapItemID));
     if (!toneMappedImageItemTmp) {
         return AVIF_RESULT_OK;
     }
 
     assert(gainMapItemID != 0);
-    avifDecoderItem * gainMapItemTmp;
-    AVIF_CHECKRES(avifMetaFindOrCreateItem(data->meta, gainMapItemID, &gainMapItemTmp));
-    if (avifDecoderItemShouldBeSkipped(gainMapItemTmp)) {
+    avifDecoderItem * gainMapItem;
```
Comment (wantehchang) [inline src/read.c:4381]: Nit: I wonder if the original code uses a `gainMapItemTmp` local variable so that the output parameter `*gainMapItem` is only assigned a value when the function returns AVIF_RESULT_OK. If so, the new code should do the same with the `*gainMapItemIndex` output parameter.

File: src/read.c [hunk 8]
```diff
@@ -4712,6 +4728,14 @@ avifResult avifDecoderReset(avifDecoder * decoder)
         }
 #endif // AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP
 
+        // When searching for alpha and gainmap items, it is possible that the data->meta->items array is re-sized in calls to
+        // avifMetaFindOrCreateItem. So it is not safe to store pointers into that array until all the items are either found or
```
Comment (wantehchang) [inline src/read.c:4732]: Nit: "that array" is not clear. Replaced it with "the mainItems array".

File: src/read.c [hunk 9]
```diff
@@ -4342,41 +4351,49 @@ static avifResult avifDecoderDataFindToneMappedImageItem(const avifDecoderData *
...
         avifDiagnosticsPrintf(data->diag, "Box[tmap] gain map item %d is not a supported image type", gainMapItemID);
         return AVIF_RESULT_INVALID_TONE_MAPPED_IMAGE;
     }
+    for (uint32_t itemIndex = 0; itemIndex < data->meta->items.count; ++itemIndex) {
+        if (gainMapItem->id == data->meta->items.item[itemIndex].id) {
+            *gainMapItemIndex = itemIndex;
+            break;
+        }
+    }
+    assert(*gainMapItemIndex >= 0);
```
Comment (wantehchang) [inline src/read.c:4393]: Nit/optional: It is unfortunate that we have to search in the `data->meta->items` array to find the item index. We can consider adding a variant of the `avifMetaFindOrCreateItem` function that outputs an item index rather than an item pointer.

**Review by wantehchang — APPROVED** [review]
LGTM. I suggest some changes. I will take a look at this CL again.

File: src/read.c [hunk 10]
```diff
@@ -4342,41 +4351,49 @@ static avifResult avifDecoderDataFindToneMappedImageItem(const avifDecoderData *
     return AVIF_RESULT_OK;
 }
 
-// Finds a 'tmap' (tone mapped image item) box associated with the given 'colorItem',
+// Finds a 'tmap' (tone mapped image item) box associated with the given colorItem at index 'colorItemIndex',
```
Comment (wantehchang) [inline src/read.c:4354]: Nit: the given colorItem => the color item

File: src/read.c (hunk 7 above)
Comment (wantehchang) [reply src/read.c:4381]: The answer to my question is Yes. Please see the comment at line 4429:

```
    // Only set the output parameters after everything has been validated.
    *toneMappedImageItem = toneMappedImageItemTmp;
    return AVIF_RESULT_OK;
```

So `*gainMapItemIndex` should receive the same treatment as `*gainMapItem` in the original code.

File: src/read.c (hunk 2 above)
Comment (vigneshvg) [reply src/read.c]: Done.

File: src/read.c (hunk 3 above)
Comment (vigneshvg) [reply src/read.c]: Done.

File: src/read.c (hunk 4 above)
Comment (vigneshvg) [reply src/read.c]: Done.

File: src/read.c (hunk 5 above)
Comment (vigneshvg) [reply src/read.c]: Done.

File: src/read.c (hunk 6 above)
Comment (vigneshvg) [reply src/read.c]: Done.

File: src/read.c (hunk 10 above)
Comment (vigneshvg) [reply src/read.c:4354]: Done.

File: src/read.c (hunk 7 above)
Comment (vigneshvg) [reply src/read.c:4381]: Done.

File: src/read.c (hunk 8 above)
Comment (vigneshvg) [reply src/read.c]: Done.

File: src/read.c (hunk 9 above)
Comment (vigneshvg) [reply src/read.c]: Ack, i thought about this as well. But since this is the only use, i will leave this as-is.

File: src/read.c [hunk 11]
```diff
@@ -4211,39 +4214,40 @@ static avifBool avifDecoderItemIsAlphaAux(avifDecoderItem * item, uint32_t color
     return auxCProp && isAlphaURN(auxCProp->u.auxC.auxType);
 }
 
-// Finds the alpha item whose parent item is colorItem and sets it in the alphaItem output parameter. Returns AVIF_RESULT_OK on
-// success. Note that *alphaItem can be NULL even if the return value is AVIF_RESULT_OK. If the colorItem is a grid and the alpha
-// item is represented as a set of auxl items to each color tile, then a fake item will be created and *isAlphaItemInInput will be
-// set to AVIF_FALSE. In this case, the alpha item merely exists to hold the locations of the alpha tile items. The data of this
-// item need not be read and the pixi property cannot be validated. Otherwise, *isAlphaItemInInput will be set to AVIF_TRUE when
-// *alphaItem is not NULL.
+// Finds the alpha item whose parent item is the color item and sets its index in the alphaItemId output parameter. Returns
```
Comment (wantehchang) [inline src/read.c:4217]: alphaItemId => alphaItemIndex

File: src/read.c [hunk 12]
```diff
@@ -4292,7 +4301,7 @@ static avifResult avifMetaFindAlphaItem(avifMeta * meta,
 }
 
 #if defined(AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP)
-// Finds a 'tmap' (tone mapped image item) box associated with the given 'colorItem'.
+// Finds a 'tmap' (tone mapped image item) box associated with the given color item.
```
Comment (wantehchang) [inline src/read.c:4304]: Optional: Undo this change (i.e., change `color item` back to `'colorItem'`) ?

File: src/read.c (hunk 10 above)
Comment (wantehchang) [reply src/read.c:4354]: The change I suggested was not made.

File: src/read.c [hunk 13]
```diff
@@ -4712,6 +4730,14 @@ avifResult avifDecoderReset(avifDecoder * decoder)
         }
 #endif // AVIF_ENABLE_EXPERIMENTAL_GAIN_MAP
 
+        // When searching for alpha and gainmap items, it is possible that the data->meta->items array is re-sized in calls to
+        // avifMetaFindOrCreateItem. So it is not safe to store pointers into data->meta->items array until all the items are
```
Comment (wantehchang) [inline src/read.c:4734]: Nit: add "the" before "data->meta->items array"

I am sorry I misunderstood what "that array" referred to. You can change this back to "that array".

**Review by wantehchang — APPROVED** [review]
LGTM. Just some comment nits.

File: src/read.c (hunk 11 above)
Comment (vigneshvg) [reply src/read.c:4217]: Done.

File: src/read.c (hunk 12 above)
Comment (vigneshvg) [reply src/read.c:4304]: Done.

File: src/read.c (hunk 10 above)
Comment (vigneshvg) [reply src/read.c:4354]: sorry i made it in the wrong place. fixed it now.

File: src/read.c (hunk 13 above)
Comment (vigneshvg) [reply src/read.c:4734]: I added `the` and left it as data->meta->items.
